### PR TITLE
feat(producer): add MaxBufferBytes to limit retry buffer size

### PR DESCRIPTION
**Pull request by wanwenli**
# Summary

This PR introduces the MaxBufferBytes configuration to Sarama’s producer, allowing users to limit the retry buffer size based on the total byte size of messages. This complements the existing MaxBufferLength configuration, which limits the number of messages in the retry buffer. The implementation ensures more robust control over memory usage and prevents potential OOM issues when dealing with large messages.

The changes are built upon the foundation laid in https://github.com/IBM/sarama/pull/3026, further enhancing the retry buffer’s configurability and usability.

# Key Changes

- Added a new Producer.Retry.MaxBufferBytes configuration.
- The type of MaxBufferBytes is int64, as the maximum value of int32 (approximately 2 GB) is insufficient for buffers holding many messages of large average size, such as 1 MB. Using int64 ensures the buffer can handle higher limits if needed.
- Ensured that any value below the functional lower limit of 32 MB is pushed to this minimum.
- Updated the retry handler to enforce both MaxBufferBytes and MaxBufferLength constraints.
- Added relevant unit tests to validate functionality for both configurations, including edge cases and scenarios with unlimited settings.

This enhancement allows users to fine-tune producer behavior, particularly when working with large message payloads or memory-constrained environments.

### CODE REVIEW

File: async_producer_test.go [hunk 1]
```diff
@@ -2199,57 +2200,86 @@ func TestPreventRetryBufferOverflow(t *testing.T) {
...
+					Value: val,
+				}
+				producer.Input() <- msg
+			}
+
+			producer.AsyncClose()
+			wg.Wait()
+
+			require.Equal(t, successes+producerErrors, tt.numMessages, "Expected all messages to be processed")
+			require.True(t, errorFound, "Expected at least one error matching ErrProducerRetryBufferOverflow")
```
Comment (puellanivis) [inline async_producer_test.go:2280]: There’s not really a reason why these should be `require` instead of `assert`. https://go.dev/wiki/TestComments#keep-going

Comment (wanwenli) [reply async_producer_test.go]: Thank you for your feedback! It helped me learn new things, and I’ve updated my code accordingly.

**Comment by wanwenli** [comment]
@dnwe can you please review this PR? Thanks!

File: async_producer.go [hunk 2]
```diff
@@ -1239,15 +1258,18 @@ func (p *asyncProducer) retryHandler() {
 		}
 
 		buf.Add(msg)
+		currentByteSize += int64(msg.ByteSize(version))
 
-		if maxBufferSize > 0 && buf.Length() >= maxBufferSize {
+		if (maxBufferLength > 0 && buf.Length() >= maxBufferLength) || (maxBufferBytes > 0 && currentByteSize >= maxBufferBytes) {
```
Comment (puellanivis) [inline async_producer.go:1263]: [improvement?] Hm… 🤔  an idea would be to maybe invert the `if`, so the body could be an early `continue`. This follows https://go.dev/wiki/CodeReviewComments#indent-error-flow

**Review by puellanivis — COMMENTED** [review]
The only thing left to comment about is a possible alternative, that may or may not be an improvement.

Otherwise, this looks like it does what it intends.

File: async_producer.go (hunk 2 above)
Comment (wanwenli) [reply async_producer.go]: thanks. I've taken your advice and inverted the `if` condition via De Morgan's laws. This improvement also reduces the indentation level.

**Review by dnwe — APPROVED** [review]
Thank you for this contribution, and for refining it based on the review feedback. The changes LGTM!
